### DISCUSSION TITLE
Drop support for Python 3.8

### DISCUSSION THREAD

**Discussion by bitnom** [microsoft/autogen#1173] [General]
Basis for this discussion is #1167 as well as the typical reasons for dropping support of an older Python version.

My thinking on the issue is that the maximum version range should be supported up until inclusion of a version begins to incur any amount of technical debt or sacrifice, which appears to be the case. It can only continue downhill from there. Instead, we should continue the climb up the mountain as soon as we have a reasonable excuse to.

**Comment by rickyloynd-microsoft** [comment]
Thanks for raising the topic and #1167, which gives details about the matplotlib version conflict. To further this discussion, does anyone know of other costs or limitations incurred by python 3.8?

**Reply by bitnom** [reply]
I posted #1188 in response to this question but I wouldn't consider it to be a blocking issue.

**Reply by davorrunje** [reply]
The reason for dropping 3.8 is matplotlib, but it is only used in notebooks and not the autogen itself. Notebook-specific dependencies should be moved from the project and declared in notebooks as they typically are. Some notebooks will only work with specific libraries and versions of Python, but that is not the reason to limit the autogen library itself. The library should have the minimum requirements possible as it should adopt to wide range of applications and their environments.

**Reply by bitnom** [reply]
that's right. I wasn't really considering that notebooks handle their own deps. We will still run into the conflict, just not in the main pyproject.toml. Ideally, pre-commit tests would check the demos, and at least throw a warning if one is about to break.

Regarding 3.8, I gathered some stats on autogen module usage. Here are the Python versions people are installing autogen on by percentage/download count:

| category | percent | downloads |
|----------|---------|-----------|
| 3.11     |  40.55% |   112,527 |
| 3.10     |  37.50% |   104,065 |
| 3.9      |   6.98% |    19,360 |
| 3.12     |   5.66% |    15,720 |
| 3.8      |   5.42% |    15,047 |
| null     |   3.84% |    10,664 |
| 3.3      |   0.03% |        75 |
| 3.13     |   0.01% |        32 |
| 2.7      |   0.00% |         7 |
| 3.7      |   0.00% |         3 |
| 3.5      |   0.00% |         2 |
| **Total**    |         | **277,502** |

15,047 / 5.42% of installs is not insignificant. I withdraw my previous position. 3.8 should probably stay until that number is significantly less.

**Reply by ekzhu** [reply]
thanks for the numbers.

**Comment by davorrunje** [comment]
AutoGen by itself doesn't have many dependencies and can easily support Python 3.8 until its eol later this year (https://devguide.python.org/versions/). It is a pain sometimes to write for 3.8, but what can you do. Different examples and notebooks will be very different because they will sometime use the latest experimental libraries and that's fine for each of them. It is highly likely they will not support older versions of Python and they will be in conflict with each other. AutoGen has a huge potential and I believe those download numbers will be huge and will have to work in many different environments. There is no need to limit that because of examples or to limit examples only to use well supported libs.

**Reply by bitnom** [reply]
it's all good. I thought it was going to be < 1%
